### ISSUE TITLE
CMake has problems finding Python on 32 bits.

### ISSUE THREAD

**Issue by OscarL** [haikuports/haikuports#11155]
As mentioned [here](https://github.com/haikuports/haikuports/pull/11141#issuecomment-2363835878), trying to build gemrb on 32 bits gives:

```
CMake Error at /boot/system/data/cmake/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find Python (missing: Python_INCLUDE_DIRS Development
  Development.Module Development.Embed) (found suitable version "3.10",
  minimum required is "3")
Call Stack (most recent call first):
  /boot/system/data/cmake/Modules/FindPackageHandleStandardArgs.cmake:600 (_FPHSA_FAILURE_MESSAGE)
  /boot/system/data/cmake/Modules/FindPython/Support.cmake:3867 (find_package_handle_standard_args)
  /boot/system/data/cmake/Modules/FindPython.cmake:574 (include)
  cmake/Helpers.cmake:134 (FIND_PACKAGE)
  CMakeLists.txt:98 (CONFIGURE_PYTHON)
```

Seems like `/system/data/cmake/Modules/FindPython/Support.cmake` only ever uses paths like `"lib/python${version}/config-${version}"`, where for 32 bits it should be "lib/x86/python${version}/config-${version}".

@jmairboeck pointed out `${CMAKE_HAIKU_SECONDARY_ARCH_SUBDIR}` (seen [here](https://github.com/Kitware/CMake/blob/ac8738c0770667086d43aba024dc749602252011/Modules/Platform/Haiku.cmake#L43)) that could be used to fix the issue.

**Comment by lonemadmax** [comment]
Can't test anything right now, so I'll probably say something stupid.

> Seems like `/system/data/cmake/Modules/FindPython/Support.cmake` only ever uses paths like `"lib/python${version}/config-${version}"`, where for 32 bits it should be "lib/x86/python${version}/config-${version}".

But it isn't? The package list says it's `lib/python3.10/config-3.10`, and there's also `develop/headers/python3.10/Python.h`. But then `develop/lib/x86/libpython3.10.so` and `lib/x86/libpython3.10.so.1.0`.

Notice also it says `found suitable version "3.10"`, and the problem seems to be missing the include dirs (and maybe something else). So maybe the problem is the opposite: it does find the library and then it can't find Python.h.

What does python return for `sysconfig.get_path('platinclude')` and `sysconfig.get_path('include')`? Those seem to be the [first search options](https://github.com/Kitware/CMake/blob/5e984bb35232116a54de7db39788cb162ca7c263/Modules/FindPython/Support.cmake#L520). If it doesn't like the result, it seems it tries [something from the lib dir](https://github.com/Kitware/CMake/blob/5e984bb35232116a54de7db39788cb162ca7c263/Modules/FindPython/Support.cmake#L3470), which, if anything, would fall to the last option for us, and if we are lucky, use `/system/develop` as a hint. But then it will look for [`include/pyhton3.10`](https://github.com/Kitware/CMake/blob/5e984bb35232116a54de7db39788cb162ca7c263/Modules/FindPython/Support.cmake#L318), but again, we have that as `headers/python3.10`.

So, to recap, I think that:
- `import sysconfig; print(sysconfig.get_path('platinclude')); print(sysconfig.get_path('include'))` does not point to the correct paths. A python expert would have to confirm this.
- Meanwhile, changing those `include/` in lines 315 and 318 of the cmake file to `headers/` might work.

**Comment by lonemadmax** [comment]
> saw it calling `python-config.py` (at least for some platforms)

Another thing to check, I guess. I did not see that one.

The scheme is a fallback. It checks both get_path(whatever) and get_path(whatever, scheme).

**Comment by lonemadmax** [comment]
I have finally been able to test it myself. Thank you cmake for your `--trace` options.

In both 32 and 64 bits, `/bin/python3.10-config` is found as the configuration program, and it gives the same (and correct) paths it's asked for. But then we have "[check that config tool match library architecture](https://github.com/Kitware/CMake/blob/5e984bb35232116a54de7db39788cb162ca7c263/Modules/FindPython/Support.cmake#L2797)".

`CMAKE_LIBRARY_ARCHITECTURE` is defined as `x86` in the 32 bits run (as it's in fact running for a secondary arch) and undefined in the 64 bits one. So for 32 bits it expects `/bin/python3.10-config --configdir` to give a path with `x86` on it. As it doesn't (it's `/boot/system/lib/python3.10/config-3.10`, and that's correct), it goes on with no config utility and ultimately tries `/boot/system/develop/include/{python*/,}/Python.h`, but we have it under `headers`, not `include`.

I guess we don't want to change `CMAKE_LIBRARY_ARCHITECTURE`, as it's used to find libraries and packages. The question then is if we want to change how we package python or keep it as we do, without the secondary arch prefix, in which case I guess we should change cmake's python finder.

Whatever the answer we probably want to change those `include` for `headers` anyway.

Notice that even with that, package creation still fails in this case in the install phase with: Failed to open output file "/packages/gemrb_x86-0.9.3-1/.self/bin/x86/gemrb": no such file or directory. But that should have nothing to do with this issue.

**Comment by lonemadmax** [comment]
> Awesome @lonemadmax! Thanks. I tried reading the whole .cmake file, but that didn't help me much :-). Thanks for the `--trace` tip!
> 
> In regards to whether fix this by moving Python to x86 or fix CMake's Python finder:
> 
> I think it might be simpler to do the latter, as the first might introduce a myriad of problems for compiled python modules (that we would have to make sure play nice with the `lib` vs `lib/x86` distinction).
> 
> Do you have a .patch you could share (even if WIP)? Would be nice to integrate it on the updated CMake 3.30.3 .recipe build yesterday on 32 bits. I guess I can just add it as "draft" PR, and you reply there whenever you can (about to run a build and test-run for that version on 64 bits before the PR).
> 
> > Notice that even with that, package creation still fails in this case in the install phase with: Failed to open output file "/packages/gemrb_x86-0.9.3-1/.self/bin/x86/gemrb": no such file or directory. But that should have nothing to do with this issue.
> 
> Probably changing `$binDir` to `$prefix/bin` might solve that, but yea... totally unrelated to this.

Simpler, but is it the right thing to do? I don't use x86 so I don't know its intricacies and can't make a call, but I would expect secondary arch stuff to be under `x86/`.

> Do you have a .patch

I don't. For testing I just changed the three ´include´ in CPython case in `_PYTHON_GET_PATH_SUFFIXES` for `headers`. That's lines 316, 318 and 320. But someone serious would also check correctness for the other implementations (if we have them ported at all), remove the arch check if we're keeping python as we have it now (which would probably make the include vs headers change unnecessary, but still) and review the whole thing for other mishaps.

Don't mistake my participation for my being a cmake expert. I can't even start a simple project with it. It's just that I like getting to where problems lie and leave the solutions to those that may do a good job with the information.

**Comment by korli** [comment]
Does this cmake version update trigger the problems on x86?

**Comment by korli** [comment]
FindPython is a bit dumb, it only uses the config when it found an interpreter. Adding the interpreter when development is searched helped here. Patch for FindPython/Support.cmake:
```
 if ("Development" IN_LIST ${_PYTHON_PREFIX}_FIND_COMPONENTS)
   list (APPEND ${_PYTHON_PREFIX}_FIND_COMPONENTS "Development.Module" "Development.Embed")
+  if (HAIKU)
+    list (APPEND ${_PYTHON_PREFIX}_FIND_COMPONENTS "Interpreter")
+  endif()
 endif()
 list (REMOVE_DUPLICATES ${_PYTHON_PREFIX}_FIND_COMPONENTS)
 foreach (_${_PYTHON_PREFIX}_COMPONENT IN ITEMS Interpreter Compiler Development Development.Module Development.SABIModule Development.Embed NumPy)

```
